services/web-server/src/PulseEngine/Subscription.js

```diff
@@ -0,0 +1,162 @@
+import Debug from 'debug';
+
+const debug = Debug('Subscription');
+
+export default class Subscription {
+  constructor({ subscriptionId, handleMessage, handleError, subscriptions }) {
+    this.subscriptionId = subscriptionId;
+    this.handleMessage = handleMessage;
+    this.handleError = handleError;
+    this.subscriptions = subscriptions;
+
+    // state tracking for reconciliation
+    this.listening = false;
+    this.unsubscribed = false;
+  }
+
+  /**
+   * Reset this subscription to its non-listening state
+   */
+  reset() {
+    this.listening = false;
+    this.connection = null;
+    this.channel = null;
+  }
+
+  /**
+   * Flag this subscription as needing unsubscription at the next
+   * reconciliation
+   */
+  unsubscribe() {
+    this.unsubscribed = true;
+  }
+
+  /**
+   * Reconcile the AMQP state with this subscription's state.
+   */
+  async reconcile(client, connection) {
+    const { subscriptionId, listening, unsubscribed } = this;
+    const queueName = client.fullObjectName('queue', subscriptionId);
+
+    if (!this.channel || connection !== this.connection) {
+      this.connection = connection;
+      this.channel = await connection.amqp.createChannel();
+      // intercept 'error' events so they don't kill the connection (they are
+      // also raised as exceptions and handled that way)
+      this.channel.on('error', () => {});
+    }
+
+    const { channel } = this;
+
+    // all errors from here on are handled by calling the subscription's `handleError`
+    // method and considering the subscription reconciled.  So errors are not bubbled
+    // up to the caller and will not interfere with other subscriptions.
+    try {
+      if (listening && unsubscribed) {
+        debug(`Unbinding subscription ${subscriptionId}`);
+        await channel.cancel(this.consumerTag);
+        await channel.deleteQueue(queueName);
+        await channel.close();
+        this.listening = false;
+      } else if (!listening && !unsubscribed) {
+        debug(`Binding subscription ${subscriptionId}`);
+        const { handleError, handleMessage, subscriptions } = this;
+
+        // declare the queue, with autoDelete and exclusive both set to false so that
+        // the queue will stick around if we need to reconnect, but with a short TTL
+        // so that if we "forget" about the queue, it will be deleted quickly.
+        await channel.assertQueue(queueName, {
+          exclusive: false,
+          durable: true,
+          autoDelete: false,
+          expires: 30000, // 30 seconds, long enough for a reconnect
```

Comment:
I wonder if this is gonna yield the same error as I got when working on the TTL bug 🤔

Reply:
I hope not -- these queues should not already exist, so the consistency error shouldn't occur.